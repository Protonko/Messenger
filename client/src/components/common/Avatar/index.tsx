import {useMemo, FC} from 'react'
import classNames from 'classnames'
import {Sizes} from 'models/common/sizes'
import {ColorGeneratorByName} from 'utils/ColorGeneratorByName'
import {isDarkColor} from 'utils/isDarkColor'
import {Loader} from 'components/common/Loader'

export interface IPropsAvatar {
  src: string
  name?: string
  size?: Sizes
  customStyles?: string
}

export const Avatar: FC<IPropsAvatar> = ({
  src,
  name,
  size = Sizes.MEDIUM,
  customStyles,
}) => {
  const backgroundColor = useMemo(
    () => new ColorGeneratorByName(name ?? '').generate(),
    [src, name],
  )
  const avatarClassname = classNames(
    'avatar',
    {'avatar--sm': size === Sizes.SMALL},
    {'avatar--lg': size === Sizes.LARGE},
    {[customStyles ?? '']: customStyles},
  )

  if (!src && name) {
    const classNamesText = classNames(
      'avatar__text',
      {'avatar__text--light': isDarkColor(backgroundColor)},
      {'avatar__text--dark': !isDarkColor(backgroundColor)},
    )

    return (
      <div className={avatarClassname} style={{backgroundColor}}>
        <span className={classNamesText}>{name.charAt(0)}</span>
      </div>
    )
  }

  return (
    <div className={avatarClassname}>
      <img className="avatar__image" src={src} alt="" />
      <Loader />
    </div>
  )
}
