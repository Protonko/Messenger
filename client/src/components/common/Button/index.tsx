import {FC, HTMLProps, ReactNode} from 'react'

interface IButtonProps extends HTMLProps<HTMLButtonElement> {
  text?: string
  icon?: ReactNode
  leftAddons?: ReactNode
  rightAddons?: ReactNode
  type?: 'button' | 'submit' | 'reset'
}

export const Button: FC<IButtonProps> = ({
  text,
  icon,
  leftAddons,
  rightAddons,
  ...buttonProps
}) => {
  const renderButtonContent = () => {
    return (
      <>
        {text && (
          <span className="button__text">
            {text}
          </span>
        )}
        {icon && (
          <span className="button__icon">
            {icon}
          </span>
        )}
      </>
    )
  }

  return (
    <button className="button" {...buttonProps}>
      {leftAddons}
      {renderButtonContent()}
      {rightAddons}
    </button>
  )
}
