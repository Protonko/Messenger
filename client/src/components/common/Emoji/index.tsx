import 'emoji-mart/css/emoji-mart.css'
import {FC, useState, useRef} from 'react'
import {EmojiData, Picker} from 'emoji-mart'
import {ReactComponent as Smile} from 'assets/icons/smile.svg'
import {useOutsideClick} from 'hooks/useOutsideClick'

export interface IPropsEmoji {
  onSelect: (emoji: EmojiData) => void
}

export const Emoji: FC<IPropsEmoji> = ({onSelect}) => {
  const [open, setOpen] = useState(false)
  const ref = useRef(null)
  const initiatorRef = useRef(null)

  useOutsideClick(ref, () => setOpen(false), initiatorRef)

  const onSelectEmoji = (emoji: EmojiData) => {
    onSelect(emoji)
    setOpen(false)
  }

  return (
    <div className="emoji">
      {open && (
        <div className="emoji__popup" ref={ref}>
          <Picker native={true} onSelect={onSelectEmoji} />
        </div>
      )}

      <Smile
        className="emoji__button"
        ref={initiatorRef}
        onClick={() => setOpen(!open)}
      />
    </div>
  )
}
