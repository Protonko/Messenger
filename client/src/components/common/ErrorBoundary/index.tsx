import {Component, ErrorInfo} from 'react';

interface IProps {}
interface IState {
  error: null | Error;
  errorInfo: null | ErrorInfo;
  hasError: boolean;
}

export class ErrorBoundary extends Component<IProps, IState> {
  constructor(props: IProps) {
    super(props);

    this.state = {
      error: null,
      errorInfo: null,
      hasError: false,
    }
  }

  static getDerivedStateFromError(error: Error) {
    console.log(error)
    return {
      hasError: true,
    };
  }

  componentDidCatch(error: Error, errorInfo: ErrorInfo) {
    this.setState({
      error: error,
      errorInfo: errorInfo
    })
  }

  _renderDetailInfo() {
    if (process.env.NODE_ENV === 'development') {
      return (
        <div>

        </div>
      )
    }

    return null
  }

  render() {
    if (this.state.hasError) {
      return (
        <>
          <h1>Что-то пошло не так.</h1>

          {process.env.NODE_ENV === 'development' && (
            <details style={{ whiteSpace: 'pre-wrap' }}>
              {this.state.error && this.state.error.toString()}
              <br />
              {this.state.errorInfo?.componentStack}
            </details>
          )}
        </>
      )
    }

    return this.props.children;
  }
}
