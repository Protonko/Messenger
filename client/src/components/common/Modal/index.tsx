import {useRef, FC, ReactPortal, ReactNode, RefObject} from 'react'
import {createPortal} from 'react-dom'
import {Transition, TransitionStatus} from 'react-transition-group'
import classNames from 'classnames'
import {useOutsideClick} from 'hooks/useOutsideClick'
import {HEIGHT_MODAL, WIDTH_MODAL, TOGGLE_MODAL_TIMEOUT} from 'static/constants'
import {rem} from 'utils/rem'

export interface IPropsModal {
  children: ReactNode
  initiatorRef?: RefObject<HTMLElement>
  modalVisibility: boolean
  toggleVisibilityModal?: (visibility: boolean) => void
  onClose?: () => void
  timeout?: number
  width?: number
  height?: number
  additionalClassName?: string
  withModalTemplate?: boolean
}

const TRANSITION_CLASSNAMES = {
  entering: 'modal--fade-in',
  entered: '',
  exiting: 'modal--fade-out',
  exited: '',
  unmounted: '',
}

export const ModalContent: FC<IPropsModal> = ({
  children,
  modalVisibility,
  toggleVisibilityModal,
  onClose,
  timeout = TOGGLE_MODAL_TIMEOUT,
  width = WIDTH_MODAL,
  height = HEIGHT_MODAL,
  additionalClassName,
  initiatorRef,
  withModalTemplate,
}) => {
  const modal = useRef<HTMLElement>(null)
  useOutsideClick<HTMLElement>(
    modal,
    () => {
      onClose?.()
      toggleVisibilityModal?.(false)
    },
    initiatorRef,
  )

  const renderContent = (state: TransitionStatus) => {
    const classNamesModal = classNames(
      'modal',
      {[TRANSITION_CLASSNAMES[state]]: !!TRANSITION_CLASSNAMES[state]},
      {'modal--blur': !withModalTemplate},
    )
    const classNamesModalContent = classNames('modal__popup', {
      [additionalClassName ?? '']: !!additionalClassName,
    })

    if (withModalTemplate) {
      return (
        <div className={classNamesModal}>
          <article
            className={classNamesModalContent}
            style={{width: rem(width), height: rem(height)}}
            ref={modal}
          >
            {children}
          </article>
        </div>
      )
    }

    return <div className={classNamesModal}>{children}</div>
  }

  return (
    <Transition
      nodeRef={modal}
      in={modalVisibility}
      timeout={timeout}
      mountOnEnter={true}
      unmountOnExit={true}
    >
      {renderContent}
    </Transition>
  )
}

export const Modal = ({children, ...props}: IPropsModal): ReactPortal => {
  const portalRoot = document.getElementById('portal-modal') as HTMLElement
  const el = useRef(document.createElement('div'))
  portalRoot.appendChild(el.current)

  return createPortal(
    <ModalContent {...props}>{children}</ModalContent>,
    el.current,
  )
}
