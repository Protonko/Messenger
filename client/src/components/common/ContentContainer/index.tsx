import type {FC} from 'react'
import {Loader} from 'components/common/Loader'
import {Text} from 'components/common/Text'

interface ContentContainerProps {
  loading: boolean
  errorMessage: string | null
  children: JSX.Element
}

export const ContentContainer: FC<ContentContainerProps> = ({
  loading,
  errorMessage,
  children,
}) => {
  if (errorMessage) {
    return <Text>{errorMessage}</Text>
  }

  if (loading) {
    return <Loader center={true} />
  }

  return children
}
