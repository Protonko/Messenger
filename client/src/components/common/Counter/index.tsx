// types
import {Status} from 'models/common/status'

import {FC} from 'react'
import classNames from 'classnames'

export interface IPropsCounter {
  count: number | '99+',
  status?: Status,
}

export const Counter: FC<IPropsCounter> = ({
  count,
  status = Status.ACTIVE,
}) => {
  const classNameStatus = classNames(
    'counter',
    {'counter--muted': status === Status.MUTED}
  )

  return (
    <div className={classNameStatus}>
      <span className="counter__text">
        {count}
      </span>
    </div>
  )
}
