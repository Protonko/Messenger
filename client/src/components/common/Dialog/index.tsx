import type {FC} from 'react'
import type {IDialog} from 'models/dialog'
import {ReadStatus} from 'models/common/status'
import {TextSize, TextTypes} from 'models/common/text'
import {ReactComponent as DoubleCheck} from 'assets/icons/double-check.svg'
import {ReactComponent as Check} from 'assets/icons/check.svg'
import {COLORS} from 'static/colors'
import {Text} from 'components/common/Text'
import {Counter} from 'components/common/Counter'
import {Avatar} from 'components/common/Avatar'

export const Dialog: FC<IDialog> = ({
  name,
  description,
  avatar,
  edited,
  createdAt,
  updatedAt,
  messages,
  status,
  readStatus,
}) => {
  const counter = messages.toString().length > 2 ? '99+' : messages

  const renderStatus = () => {
    if (!!counter) {
      return <Counter count={counter} status={status}/>
    }

    if (readStatus) {
      const props = {
        width: 15,
        height: 15,
        color: COLORS.dustyGray,
      }
      return readStatus === ReadStatus.READ ? <DoubleCheck {...props} /> : <Check {...props} />
    }

    return null
  }

  return (
    <div className="dialog">
      <Avatar src={avatar} />

      <div className="dialog__data">
        <div className="dialog__text">
          <Text
            type={TextTypes.h4}
            customStyles="dialog__text-title"
            numberOfLines={1}
          >
            {name}
          </Text>
          <Text>{description}</Text>
        </div>

        <div className="dialog__info">
          <Text
            type={TextTypes.mixed}
            size={TextSize.EXTRA_SMALL}
            customStyles="dialog__info-date"
          >
            {new Date(edited ? updatedAt : createdAt).toLocaleDateString()}
          </Text>

          {renderStatus()}
        </div>
      </div>
    </div>
  )
}
