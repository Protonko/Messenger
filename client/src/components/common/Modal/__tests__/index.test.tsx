import {mount, ReactWrapper} from 'enzyme'
import {Modal, ModalContent, IPropsModal} from 'components/common/Modal'
import {rem} from 'utils/rem'
import {Transition} from 'react-transition-group'

const ELEMENT_SELECTORS = {
  modal: 'modal',
  popup: 'modal__popup',
  customStyle: 'modal__test',
}

describe('Modal component', () => {
  const Child = () => <div>Yolo</div>
  const portalRoot = document.createElement('div')
  const body = document.querySelector('body')
  const fn = jest.fn()

  portalRoot.setAttribute('id', 'portal-modal')
  body!.appendChild(portalRoot)

  let props: IPropsModal
  let component: ReactWrapper<IPropsModal>

  beforeEach(() => {
    props = {
      children: Child,
      modalVisibility: true,
      toggleVisibilityModal: fn,
      withModalTemplate: true,
    }

    component = mount(
      <Modal {...props}>
        <Child />
      </Modal>,
    )
  })

  it('Should create portal with ModalContent component', () => {
    expect(component.find(ModalContent).length).toBe(1)
  })

  it('Should render ModalContent component', () => {
    expect(component.find(`.${ELEMENT_SELECTORS.modal}`).length).toBe(1)
  })

  it('Match snapshot', () => {
    expect(component).toMatchSnapshot('Modal')
  })

  it('Should style properties be 10rem', () => {
    props.width = 160
    props.height = 160

    component = mount(
      <ModalContent {...props}>
        <Child />
      </ModalContent>,
    )
    const popup = component.find(`.${ELEMENT_SELECTORS.popup}`)

    expect(popup.prop('style')).toHaveProperty('width', rem(props.width))
    expect(popup.prop('style')).toHaveProperty('height', rem(props.height))
  })

  it('Should render modal without template', () => {
    props.withModalTemplate = false

    component = mount(
      <ModalContent {...props}>
        <Child />
      </ModalContent>,
    )

    expect(component).toMatchSnapshot('Modal without template')
  })

  it('Should change timeout from default value to 160', () => {
    props.timeout = 160

    component = mount(
      <ModalContent {...props}>
        <Child />
      </ModalContent>,
    )

    expect(component.find(Transition).prop('timeout')).toBe(160)
  })

  it('Modal with customStyles', () => {
    props.additionalClassName = ELEMENT_SELECTORS.customStyle

    component = mount(
      <ModalContent {...props}>
        <Child />
      </ModalContent>,
    )

    expect(component.find(`.${ELEMENT_SELECTORS.customStyle}`).length).toBe(1)
  })
})
