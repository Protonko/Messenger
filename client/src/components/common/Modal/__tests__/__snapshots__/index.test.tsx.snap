// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component Match snapshot 1`] = `
<ModalContent
  modalVisibility={true}
  toggleVisibilityModal={[MockFunction]}
>
  <Transition
    appear={false}
    enter={true}
    exit={true}
    in={true}
    mountOnEnter={true}
    nodeRef={
      Object {
        "current": <article
          class="modal__popup"
          style="width: 47rem; height: 31.25rem;"
        >
          <div>
            Yolo
          </div>
        </article>,
      }
    }
    onEnter={[Function]}
    onEntered={[Function]}
    onEntering={[Function]}
    onExit={[Function]}
    onExited={[Function]}
    onExiting={[Function]}
    timeout={400}
    unmountOnExit={true}
  >
    <div
      className="modal"
    >
      <article
        className="modal__popup"
        style={
          Object {
            "height": "31.25rem",
            "width": "47rem",
          }
        }
      >
        <Child>
          <div>
            Yolo
          </div>
        </Child>
      </article>
    </div>
  </Transition>
</ModalContent>
`;
